src/MarianInterface.cpp

```diff
@@ -119,26 +119,37 @@ MarianInterface::MarianInterface(QObject *parent)
 
                         marian::bergamot::ResponseOptions options;
                         options.alignment = true;
-                        
-                        // Using promise for a translation, and future for waiting
-                        // for that translation to turn the async service into
-                        // a blocking request.
-                        std::promise<marian::bergamot::Response> response;
-                        auto future = response.get_future();
+
+                        // Prepare our translation promise. We use this to wait on
+                        // either the translation to finish, or be cancelled. Hence
+                        // using member variable pendingTranslation_ to make it
+                        // accessible from outside the worker.
+                        {
+                            QMutexLocker locker(&lock_); // brief lock because changing member variable
+                            pendingTranslation_ = std::make_shared<Promise<Translation>>();
```

Comment:
Is this because any worker could take that translation?

Reply:
no, the workers should inside the translation service, I don't quite understand this?

Reply:
Async translation submissions from the GUI?

Reply:
MarianInterface has a single worker internally, but that one will block during translation. So the CancelablePromise is exposed to the part of MarianInterface that talks to the main thread (the Qt GUI thread) so that it can interrupt the worker.

I'm briefly locking here to have exclusive access to `pendingTranslation_`. It's a bit vague what happens when you'd trigger the old's promise cancel() just before this section. Then that interruption would be ignored I suppose…

Reply:
> Async translation submissions from the GUI?

I want that, but I don't want the backlog. Say you're typing and you just keep submitting stuff:
```
I                    (requested)
I want               (requested)
I want cake          (requested)
I want cake and      (requested)
                     (translation of "I" appears)
I want cake and eat  (requested)
                     (translation of "I want" appears)
```

With the current construction you have:
```
I                    (requested)
I want               (replaced)
I want cake          (replaced)
I want cake and      (requested)
                     (translation of "I" appears)
I want cake and eat  (requested)
                     (translation of "I want cake and" appears)
```

It's a bit like double buffering and skipping frames instead of slowing the whole game down because you can't match the 30hz refresh rate.

You might be able to get to this point with a light promise wrapper around bergamot-translator (to turn the callback into promises, sorry Jerin!) but you'd still have a bit of a mess around the logic of when to load a model (off the main thread otherwise that blocks for a second!), when to submit a translation, when to queue but not submit a translation, etc. MarianInterface effectively is there now to hide that mess from the rest of TranslateLocally.

Reply:
Yeah, we should only ever submit one thing to the translator, as we do it now. I just wonder if we can somehow simplify the logic because that while loop is on the way to being unmaintainable.... BUt let's add the functionality we want first and then we can take a look at how the logic can be simplified.

Reply:
I've reworked it a bit to use condition_variable instead. No need for a promise this way, and it is much more clear what conditions would cause it to continue the loop.

The loop is still broken into a "swap pointers with internal pointers so I can release the lock" and "do actual work" section. If you know a good way to simplify that part, I'm very interested!